components/raftstore-v2/src/operation/command/admin/merge/mod.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+pub mod prepare;
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::{
+    raft_cmdpb::{AdminCmdType, RaftCmdRequest},
+    raft_serverpb::{MergeState, PeerState, RegionLocalState},
+};
+use raft::{ProgressState, INVALID_INDEX};
+use raftstore::Result;
+use slog::{info, warn, Logger};
+use tikv_util::box_err;
+
+use crate::raft::Peer;
+
+pub enum PrepareStatus {
+    /// When a fence <idx, cmd> is present, we (1) delay the PrepareMerge
+    /// command `cmd` until all writes before `idx` are applied (2) reject all
+    /// in-coming write proposals.
+    /// Before proposing `PrepareMerge`, we first serialize and propose the lock
+    /// table. Locks marked as deleted (but not removed yet) will be
+    /// serialized as normal locks.
+    /// Thanks to the fence, we can ensure at the time of lock transfer, locks
+    /// are either removed (when applying logs) or won't be removed before
+    /// merge (the proposals to remove them are rejected).
+    ///
+    /// The request can be `None` because we needs to take it out to redo the
+    /// propose. In the meantime the fence is needed to bypass the check.
+    WaitForFence(u64, Option<RaftCmdRequest>),
+    /// In this state, all write proposals except for `RollbackMerge` will be
+    /// rejected.
+    Applied(MergeState),
+}
+
+#[derive(Default)]
+pub struct MergeContext {
+    pub prepare_status: Option<PrepareStatus>,
```

Comment:
> Seems it can never be None.

It is taken before proposing the pessimistic lock. If not an `Option` we'll need to free the `MergeContext` which will be used shortly after.

Reply:
`Option` is needed to default initialize the context. Commit merge will use it as `None`.

Reply:
How? From #13818, it will have pending_merge_result set.

Reply:
`pending_merge_result` is used at target peer, after target peer state is persisted, a message is sent to destroy source peer.